File: context.go
package gam

import "fmt"
import "github.com/emirpasic/gods/sets/hashset"
import "github.com/emirpasic/gods/stacks/linkedliststack"

type Context interface {
	Watch(ActorRef)
	Unwatch(ActorRef)
	Message() interface{}
	Become(Receive)
	BecomeStacked(Receive)
	UnbecomeStacked()
	Self() ActorRef
	Parent() ActorRef
	ActorOf(Properties) ActorRef
	Spawn(Properties) PID
	SpawnTemplate(Actor) PID
	SpawnFunc(ActorProducer) PID
	Children() []ActorRef
}

type ContextValue struct {
	*ActorCell
	message interface{}
}

func (context *ContextValue) Message() interface{} {
	return context.message
}

func NewContext(cell *ActorCell, message interface{}) Context {
	res := &ContextValue{
		ActorCell: cell,
		message:   message,
	}
	return res
}

type ActorCell struct {
	parent     ActorRef
	self       *LocalActorRef
	actor      Actor
	props      Properties
	supervisor SupervisionStrategy
	behavior   *linkedliststack.Stack
	children   *hashset.Set
	watchers   *hashset.Set
	watching   *hashset.Set
	stopping   bool
}

func (cell *ActorCell) Children() []ActorRef {
	values := cell.children.Values()
	children := make([]ActorRef, len(values))
	for i, child := range values {
		children[i] = child.(ActorRef)
	}
	return children
}

func (cell *ActorCell) Self() ActorRef {
	return cell.self
}

func (cell *ActorCell) Parent() ActorRef {
	return cell.parent
}

func NewActorCell(props Properties, parent ActorRef) *ActorCell {

	cell := ActorCell{
		parent:     parent,
		props:      props,
		supervisor: props.Supervisor(),
		behavior:   linkedliststack.New(),
		children:   hashset.New(),
		watchers:   hashset.New(),
		watching:   hashset.New(),
	}
	cell.incarnateActor()
	return &cell
}

func (cell *ActorCell) incarnateActor() {
	actor := cell.props.ProduceActor()
	cell.actor = actor
	cell.Become(actor.Receive)
}

func (cell *ActorCell) invokeSystemMessage(message SystemMessage) {
	switch msg := message.(interface{}).(type) {
	default:
		fmt.Printf("Unknown system message %T", msg)
	case *stop:
		cell.handleStop(msg)
	case *otherStopped:
		cell.handleOtherStopped(msg)
	case *watch:
		cell.watchers.Add(msg.Watcher)
	case *unwatch:
		cell.watchers.Remove(msg.Watcher)
	case *failure:
		cell.handleFailure(msg)
	case *restart:
		cell.handleRestart(msg)
	case *resume:
		cell.self.Resume()
	}
}

func (cell *ActorCell) handleStop(msg *stop) {
	cell.stopping = true
	cell.invokeUserMessage(Stopping{})
	for _, child := range cell.children.Values() {
		child.(ActorRef).Stop()
	}
	cell.tryRestartOrTerminate()
}

func (cell *ActorCell) handleOtherStopped(msg *otherStopped) {
	cell.children.Remove(msg.Who)
	cell.watching.Remove(msg.Who)
	cell.tryRestartOrTerminate()
}

func (cell *ActorCell) handleFailure(msg *failure) {
	directive := cell.supervisor.Handle(msg.Who, msg.Reason)
	switch directive {
	case ResumeDirective:
		//resume the fialing child
		msg.Who.SendSystemMessage(&resume{})
	case RestartDirective:
		//restart the failing child
		msg.Who.SendSystemMessage(&restart{})
	case StopDirective:
		//stop the failing child
		msg.Who.Stop()
	case EscalateDirective:
		//send failure to parent
		cell.parent.SendSystemMessage(msg)
	}
}

func (cell *ActorCell) handleRestart(msg *restart) {
	cell.stopping = false
	cell.invokeUserMessage(Restarting{}) //TODO: change to restarting
	for _, child := range cell.children.Values() {
		child.(ActorRef).Stop()
	}
	cell.tryRestartOrTerminate()
}

func (cell *ActorCell) tryRestartOrTerminate() {
	if !cell.children.Empty() {
		return
	}

	if !cell.stopping {
		cell.incarnateActor()
		cell.invokeUserMessage(Started{})
		return
	}

	cell.invokeUserMessage(Stopped{})
	otherStopped := &otherStopped{Who: cell.self}
	for _, watcher := range cell.watchers.Values() {
		watcher.(ActorRef).SendSystemMessage(otherStopped)
	}
}

func (cell *ActorCell) invokeUserMessage(message interface{}) {
	defer func() {
		if r := recover(); r != nil {
			failure := &failure{Reason: r, Who: cell.self}
			if cell.parent == nil {
				handleRootFailure(failure, defaultSupervisionStrategy)
			} else {
				cell.self.Suspend()
				cell.parent.SendSystemMessage(failure)
			}
		}
	}()
	behavior, _ := cell.behavior.Peek()
	behavior.(Receive)(NewContext(cell, message))
}

func (cell *ActorCell) Become(behavior Receive) {
	cell.behavior.Clear()
	cell.behavior.Push(behavior)
}

func (cell *ActorCell) BecomeStacked(behavior Receive) {
	cell.behavior.Push(behavior)
}

func (cell *ActorCell) UnbecomeStacked() {
	if cell.behavior.Size() <= 1 {
		panic("Can not unbecome actor base behavior")
	}
	cell.behavior.Pop()
}

func (cell *ActorCell) Watch(who ActorRef) {
	who.SendSystemMessage(&watch{
		Watcher: cell.self,
	})
	cell.watching.Add(who)
}

func (cell *ActorCell) Unwatch(who ActorRef) {
	who.SendSystemMessage(&unwatch{
		Watcher: cell.self,
	})
	cell.watching.Remove(who)
}

func (cell *ActorCell) ActorOf(props Properties) ActorRef {
	ref, _ := spawnChild(props, cell.self)
	cell.children.Add(ref)
	cell.Watch(ref)
	return ref
}

func (cell *ActorCell) Spawn(props Properties) PID {
	ref, pid := spawnChild(props, cell.self)
	cell.children.Add(ref)
	cell.Watch(ref)
	return pid
}

func (cell *ActorCell) SpawnTemplate(template Actor) PID {
	producer := func() Actor {
		return template
	}
	props := Props(producer)
	ref, pid := spawnChild(props, cell.self)
	cell.children.Add(ref)
	cell.Watch(ref)
	return pid
}

func (cell *ActorCell) SpawnFunc(producer ActorProducer) PID {
	props := Props(producer)
	ref, pid := spawnChild(props, cell.self)
	cell.children.Add(ref)
	cell.Watch(ref)
	return pid
}

func handleRootFailure(msg *failure, supervisor SupervisionStrategy) {
	directive := supervisor.Handle(msg.Who, msg.Reason)
	switch directive {
	case ResumeDirective:
		//resume the fialing child
		msg.Who.SendSystemMessage(&resume{})
	case RestartDirective:
		//restart the failing child
		msg.Who.SendSystemMessage(&restart{})
	case StopDirective:
		//stop the failing child
		msg.Who.Stop()
	case EscalateDirective:
		//send failure to parent
		panic("Can not escalate root level failures")
	}
}

File: languages/csharp/src/GAM/Context.cs
// -----------------------------------------------------------------------
//  <copyright file="Context.cs" company="Asynkron HB">
//      Copyright (C) 2015-2016 Asynkron HB All rights reserved
//  </copyright>
// -----------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace GAM
{
    public interface IMessageInvoker
    {
        void InvokeSystemMessage(SystemMessage msg);
        Task InvokeUserMessageAsync(object msg);
    }

    public interface IContext
    {
        PID Parent { get; }
        PID Self { get; }

        Props Props { get; }
        object Message { get; }
        PID Sender { get; }

        void Respond(object msg);
        PID[] Children();

        void Stash();
        Task NextAsync();
        PID Spawn(Props props);

        void Become(ReceiveAsync receive);
        void BecomeStacked(ReceiveAsync receive);
        void UnbecomeStacked();
    }

    public class Context : IMessageInvoker, IContext
    {
        private IActor _actor;
        private HashSet<PID> _children;
        private bool _isRestarting;
        private bool _isStopping;
        private object _message;
        private int _receiveIndex;
        private ReceiveAsync[] _receivePlugins;
        private bool _restarting;
        private Stack<object> _stash;
        private bool _stopping;
        private SupervisionStrategy _supervisionStrategy;
        private HashSet<PID> _watchers;
        private HashSet<PID> _watching;
        private Stack<ReceiveAsync> _behavior;

        public Context(Props props, PID parent)
        {
            Parent = parent;
            Props = props;
            _receivePlugins = new ReceiveAsync[] {};
            _watchers = null;
            _watching = null;
            Message = null;
            _actor = props.Producer();
            _behavior = new Stack<ReceiveAsync>();
            _behavior.Push(ActorReceiveAsync);
        }

        public PID[] Children()
        {
            return _children.ToArray();
        }

        public PID Parent { get; }
        public PID Self { get; internal set; }
        public Props Props { get; }

        public object Message
        {
            get
            {
                var r = _message as Request;
                return r != null ? r.Message : _message;
            }
            private set { _message = value; }
        }

        public PID Sender => (_message as Request)?.Sender;

        public void Stash()
        {
            if (_stash == null)
            {
                _stash = new Stack<object>();
            }
            _stash.Push(Message);
        }

        public Task NextAsync()
        {
            ReceiveAsync receive;
            if (_receiveIndex < _receivePlugins.Length)
            {
                receive = _receivePlugins[_receiveIndex];
                _receiveIndex++;
            }
            else
            {
                var func = _behavior.Peek();
                receive = func;
            }

            return receive(this);
        }

        public void Respond(object msg)
        {
            Sender.Tell(msg);
        }

        public PID Spawn(Props props)
        {
            var id = ProcessRegistry.Instance.GetAutoId();

            return SpawnNamed(props, id);
        }

        public void Become(ReceiveAsync receive)
        {
            _behavior.Clear();
            _behavior.Push(receive);
        }

        public void BecomeStacked(ReceiveAsync receive)
        {
            _behavior.Push(receive);
        }

        public void UnbecomeStacked()
        {
            if(_behavior.Count <= 1)
                throw new Exception("Can not unbecome actor base behaviour");
            _behavior.Pop();
        }

        public void InvokeSystemMessage(SystemMessage msg)
        {
            if (msg is Stop)
            {
                HandleStop();
            }
        }

        public async Task InvokeUserMessageAsync(object msg)
        {
            try
            {
                _receiveIndex = 0;
                Message = msg;

                await NextAsync();
            }
            catch (Exception x)
            {
                if (Parent == null)
                {
                }
                else
                {
                    Self.SendSystemMessage(new SuspendMailbox());
                }
                //handle supervision
            }
        }

        private void HandleStop()
        {
            _isRestarting = false;
            _isStopping = true;
            //this is intentional
            InvokeUserMessageAsync(Stopping.Instance).Wait();
            if (_children != null)
            {
                foreach (var child in _children)
                {
                    child.Stop();
                }
            }
            TryRestartOrTerminate();
        }

        private void TryRestartOrTerminate()
        {
            if (_children != null)
            {
                if (_children.Count > 0)
                {
                    return;
                }
            }

            if (_restarting)
            {
                Restart();
                return;
            }

            if (_stopping)
            {
                Stopped();
            }
        }

        private void Stopped()
        {
            ProcessRegistry.Instance.Remove(Self);
            //This is intentional
            InvokeUserMessageAsync(GAM.Stopped.Instance).Wait();
            //Notify watchers
        }

        private void Restart()
        {
            throw new NotImplementedException();
        }

        private Task ActorReceiveAsync(IContext ctx)
        {
            return _actor.ReceiveAsync(ctx);
        }

        public PID SpawnNamed(Props props, string name)
        {
            string fullname;
            if (Parent != null)
            {
                fullname = Parent.Id + "/" + name;
            }
            else
            {
                fullname = name;
            }

            var pid = Actor.InternalSpawn(props, fullname, Self);
            if (_children == null)
            {
                _children = new HashSet<PID>();
            }
            _children.Add(pid);
            Watch(pid);
            return pid;
        }

        private void Watch(PID who)
        {
            who.SendSystemMessage(new Watch(Self));
        }
    }
}